### PR TITLE
RELATED: ONE-4027 New onDrill and headers in its payload

### PR DESCRIPTION
**Pull request by ivan-nejezchleb**
New onDrill event with new payload. Intersection is now defined by Header structures from execution for all charts and contains both row and column attribues from PivotTable.

---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [ ] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [ ] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [ ] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [ ] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [ ] Successful `extended test - examples`
- [ ] Successful `extended test - storybook`
- [ ] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/2526
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2340

# Related Jira tasks

### CODE REVIEW

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4989/ : FAILURE in 12m 13s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/335/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4991/ : FAILURE in 9m 54s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/336/ : SUCCESS in 14s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/4992/ : FAILURE in 13m 09s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/337/ : SUCCESS in 7s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5003/ : FAILURE in 7m 24s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/353/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5011/ : FAILURE in 8m 50s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/368/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5016/ : SUCCESS in 10m 49s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/378/ : SUCCESS in 9s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5033/ : SUCCESS in 7m 22s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/397/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5038/ : SUCCESS in 6m 16s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/402/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5047/ : SUCCESS in 7m 39s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/414/ : SUCCESS in 6s

File: src/components/visualizations/utils/drilldownEventing.ts [hunk 1]
```diff
@@ -63,23 +80,26 @@ function fireEvent(onFiredDrillEvent: OnFiredDrillEvent, data: any, target: Even
     }
 }
 
+const getDrillPoint = (chartType: ChartType) => (point: IHighchartsPointObject): IDrillPointExtended => {
+    const customProps: Partial<IDrillPointBase> = isComboChart(chartType)
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/visualizations/utils/drilldownEventing.ts]: it is old code, just moved into sep. function

File: src/components/visualizations/utils/drilldownEventing.ts [hunk 2]
```diff
@@ -124,17 +144,62 @@ function composeDrillContextPoint(
...
+        if (isDrillIntersectionAttributeItem(header)) {
+            const { uri: itemUri, name } = header.attributeHeaderItem;
+            const { uri, identifier } = header.attributeHeader;
+            return createDrillIntersectionElement(
+                getAttributeElementIdFromAttributeElementUri(itemUri),
+                name,
+                uri,
+                identifier,
+            );
+        } else if (isMappingMeasureHeaderItem(header)) {
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/visualizations/utils/drilldownEventing.ts]: will fix it

File: src/components/core/pivotTable/agGridDrilling.ts [hunk 3]
```diff
@@ -43,10 +48,48 @@ export const getDrillRowData = (leafColumnDefs: ColDef[], rowData: { [key: strin
...
+            index: number,
+        ) => {
+            if (shouldSkipRowAttributes(index)) {
+                // skip new row attributes in legacy intersection
+                return drillItems;
+            }
+            if (isDrillIntersectionAttributeItem(intersectionElement.header)) {
+                const { attributeHeader, attributeHeaderItem } = intersectionElement.header;
+                if (drillingOnMeasure) {
+                    // different item/attribute order for drilling on measure and drilling on attr :(
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/core/pivotTable/agGridDrilling.ts]: :(

File: src/components/visualizations/headline/HeadlineTransformation.tsx [hunk 4]
```diff
@@ -76,10 +85,30 @@ class HeadlineTransformation extends React.Component<IHeadlineTransformationProp
...
+        const { onFiredDrillEvent, executionRequest, executionResponse, onDrill } = this.props;
+        const drillEventDataExtended: IDrillEventExtended = buildDrillEventData(
+            item,
+            executionRequest,
+            executionResponse,
+        );
 
-        fireDrillEvent(onFiredDrillEvent, drillEventData, target);
+        onDrill(drillEventDataExtended);
+        // old drill eventing
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/visualizations/headline/HeadlineTransformation.tsx]: fill fix it

File: src/components/core/pivotTable/agGridDrilling.ts [hunk 5]
```diff
@@ -43,10 +48,48 @@ export const getDrillRowData = (leafColumnDefs: ColDef[], rowData: { [key: strin
     }, []);
 };
 
-export const getDrillIntersection = (
-    drillItems: IMappingHeader[],
+const isDrillingOnMeasure = (measureIndex: number) => measureIndex !== -1;
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/core/pivotTable/agGridDrilling.ts]: will add comment describing structure of intersection

File: src/components/core/pivotTable/agGridDrilling.ts [hunk 6]
```diff
@@ -43,10 +48,48 @@ export const getDrillRowData = (leafColumnDefs: ColDef[], rowData: { [key: strin
...
+    );
+    const drillingOnMeasure = isDrillingOnMeasure(measureIndex);
+    const shouldSkipRowAttributes = skipRowAttributes(measureIndex);
+    const drillItems: IMappingHeader[] = intersectionExtended.reduce(
+        (
+            drillItems: IMappingHeader[],
+            intersectionElement: IDrillEventIntersectionElementExtended,
+            index: number,
+        ) => {
+            if (shouldSkipRowAttributes(index)) {
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/core/pivotTable/agGridDrilling.ts]: it would not help

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5048/ : SUCCESS in 19m 40s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/416/ : SUCCESS in 9s

**Comment by gdgate** [comment]
Build in pipeline check aborted.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5051/ : SUCCESS in 4m 19s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/419/ : SUCCESS in 6s

**Comment by ivan-nejezchleb** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5053/ : SUCCESS in 7m 59s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/424/ : SUCCESS in 10s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2646/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/425/ : SUCCESS in 10s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2058/ : UNSTABLE in 11m 18s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5054/ : SUCCESS in 10m 10s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/426/ : SUCCESS in 7s

**Comment by ivan-nejezchleb** [comment]
extended test - storybook

File: src/components/core/PivotTable.tsx [hunk 7]
```diff
@@ -465,9 +470,45 @@ export class PivotTableInner extends BaseVisualization<IPivotTableInnerProps, IP
         this.updateStickyRow();
     };
 
+    private getAttributeHeader(colId: string, columnDefs: IGridHeader[]): Execution.IAttributeHeader {
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/core/PivotTable.tsx]: fixed

File: src/components/core/PivotTable.tsx [hunk 8]
```diff
@@ -465,9 +470,45 @@ export class PivotTableInner extends BaseVisualization<IPivotTableInnerProps, IP
...
+        }, []);
+    };
+
+    private getAttributeDrillItemsForMeasureDrill = (
+        cellEvent: IGridCellEvent,
+        columnDefs: IGridHeader[],
+    ): IMappingHeader[] => {
+        const rowDrillItems = get(cellEvent, ["data", "headerItemMap"]);
+        const attributeHeaders = this.getItemAndAttributeHeaders(rowDrillItems, columnDefs);
+        return [...attributeHeaders];
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/core/PivotTable.tsx]: fixed

File: src/components/core/PivotTable.tsx [hunk 9]
```diff
@@ -500,16 +541,37 @@ export class PivotTableInner extends BaseVisualization<IPivotTableInnerProps, IP
...
+            executionContext: afm,
+            drillContext: drillContextExtended,
+        };
+
+        if (onDrill) {
+            onDrill(drillEventExtended);
+        }
+
+        // Old drill event for backward compatibility
+        const drillContext: IDrillEventContextTable = {
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/core/PivotTable.tsx]: fixed

File: src/components/core/PivotTable.tsx [hunk 10]
```diff
@@ -500,16 +541,37 @@ export class PivotTableInner extends BaseVisualization<IPivotTableInnerProps, IP
...
+            type: VisualizationTypes.TABLE,
+            element: "cell",
+            columnIndex,
+            rowIndex,
+            row,
+            intersection: convertDrillIntersectionToLegacy(intersection, afm),
+        };
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/core/PivotTable.tsx]: did my best in fix

File: src/components/visualizations/chart/chartOptions/test/extendedStackingChartOptions.spec.ts [hunk 11]
```diff
@@ -222,24 +218,18 @@ describe("getDrillableSeriesWithParentAttribute", () => {
...
             },
         },
         {
-            id: "1226",
-            title: "Direct Sales",
             header: {
-                uri: "/gdc/md/jroecoqa7jywstxy1hxp8lwl2c4nc10t/obj/1027",
-                identifier: "label.owner.department",
+                attributeHeader: dataSet.executionResponse.dimensions[1].headers[0].attributeHeader,
+                attributeHeaderItem: dataSet.executionResult.headerItems[1][0][0].attributeHeaderItem,
```
[the remark this answers is not in the document]
Comment (ivan-nejezchleb) [reply src/components/visualizations/chart/chartOptions/test/extendedStackingChartOptions.spec.ts]: drilling intersections on series points is internal structure which now has been changed. To new drill event it goes almost 1:1 and for old one it is converted to original shape

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5062/ : SUCCESS in 9m 21s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/437/ : SUCCESS in 6s

File: src/components/core/PivotTable.tsx [hunk 12]
```diff
@@ -500,31 +564,57 @@ export class PivotTableInner extends BaseVisualization<IPivotTableInnerProps, IP
...
+        return this.handleLegacyOnFireDrillEvent(drillEventExtended, cellEvent);
+    };
+
+    private handleLegacyOnFireDrillEvent = (
+        drillEventExtended: IDrillEventExtended,
+        cellEvent: IGridCellEvent,
+    ): boolean => {
+        const { onFiredDrillEvent } = this.props;
+        const { executionContext, drillContext } = drillEventExtended;
+        if (isDrillEventContextTableExtended(drillContext)) {
```
Comment (xMort) [inline src/components/core/PivotTable.tsx]: Isn't this always `true` or am I confused?

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5063/ : SUCCESS in 8m 03s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/439/ : SUCCESS in 8s

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5068/ : FAILURE in 4m 39s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/445/ : SUCCESS in 6s

**Comment by gdgate** [comment]
Build failed (check pipeline).

- gooddata-react-components-unit-tests-zuul-docker-el7 : SKIPPED
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/446/ : Tests on opensource repositories should be triggered by users in `gooddata` organization in 3s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5069/ : SUCCESS in 6m 15s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/450/ : SUCCESS in 6s

**Comment by ivan-nejezchleb** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5099/ : SUCCESS in 5m 56s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/499/ : SUCCESS in 6s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2656/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/500/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2087/ : SUCCESS in 14m 45s

**Comment by ivan-nejezchleb** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1209.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build failed (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/502/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2089/ : FAILURE in 23m 30s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5103/ : SUCCESS in 5m 05s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/507/ : SUCCESS in 6s

**Comment by ivan-nejezchleb** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1209.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/510/ : SUCCESS in 7s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/2094/ : SUCCESS in 17m 44s

**Comment by xMort** [comment]
LGTM

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/2663/ : SUCCESS in 16m 54s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/5106/ : SUCCESS in 9m 14s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1256/ : SUCCESS in 22s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/1695/ : SUCCESS in 18m 37s

**Pull request merged**